internal/generate/clusterserviceversion/clusterserviceversion_test.go

```diff
@@ -0,0 +1,379 @@
+)
+
+var _ = BeforeSuite(func() {
+	col = &collector.Manifests{}
+	Expect(col.UpdateFromDirs(goManifestRootDir, goCRDsDir)).ToNot(HaveOccurred())
+
+	cfg = readConfigHelper(goTestDataDir)
+
+	initTestCSVsHelper()
+})
+
+var _ = Describe("Generating a ClusterServiceVersion", func() {
+	format.TruncatedDiff = true
+	format.UseStringerRepresentation = true
+
+	var (
+		g            Generator
+		buf          *bytes.Buffer
+		operatorName = "memcached-operator"
+		operatorType = projutil.OperatorTypeGo
+		version      = "0.0.1"
+	)
+
+	BeforeEach(func() {
+		buf = &bytes.Buffer{}
+	})
+
+	Describe("for the new Go project layout", func() {
+
+		Context("with correct Options", func() {
+
+			var (
+				tmp string
+				err error
+			)
+
+			BeforeEach(func() {
+				tmp, err = ioutil.TempDir(".", "")
+				ExpectWithOffset(1, err).ToNot(HaveOccurred())
+			})
+
+			AfterEach(func() {
+				if tmp != "" {
+					os.RemoveAll(tmp)
+				}
+			})
+
+			It("should write a ClusterServiceVersion manifest to an io.Writer", func() {
+				g = Generator{
+					OperatorName: operatorName,
+					OperatorType: operatorType,
+					Version:      version,
+					Collector:    col,
+				}
+				opts := []Option{
+					WithBase(csvBasesDir, goAPIsDir),
+					WithWriter(buf),
+				}
+				Expect(g.Generate(cfg, opts...)).ToNot(HaveOccurred())
+				Expect(buf.String()).To(MatchYAML(newCSVStr))
+			})
+			It("should write a ClusterServiceVersion manifest to a base file", func() {
+				g = Generator{
+					OperatorName: operatorName,
+					OperatorType: operatorType,
+				}
+				opts := []Option{
+					WithBase(csvBasesDir, goAPIsDir),
+					WithBaseWriter(tmp),
+				}
+				Expect(g.Generate(cfg, opts...)).ToNot(HaveOccurred())
+				outputFile := filepath.Join(tmp, "bases", makeCSVFileName("memcached-operator"))
+				Expect(outputFile).To(BeAnExistingFile())
+				Expect(string(readFileHelper(outputFile))).To(MatchYAML(baseCSVUIMetaStr))
+			})
+			It("should write a ClusterServiceVersion manifest to a bundle file", func() {
+				g = Generator{
+					OperatorName: operatorName,
+					OperatorType: operatorType,
+					Version:      version,
+					Collector:    col,
+				}
+				opts := []Option{
+					WithBase(csvBasesDir, goAPIsDir),
+					WithBundleWriter(tmp),
+				}
+				Expect(g.Generate(cfg, opts...)).ToNot(HaveOccurred())
+				outputFile := filepath.Join(tmp, bundle.ManifestsDir, makeCSVFileName("memcached-operator"))
+				Expect(outputFile).To(BeAnExistingFile())
+				Expect(string(readFileHelper(outputFile))).To(MatchYAML(newCSVStr))
+			})
+		})
+
+		Context("with incorrect Options", func() {
+
+			BeforeEach(func() {
+				g = Generator{
+					OperatorName: operatorName,
+					OperatorType: operatorType,
+					Version:      version,
+					Collector:    col,
+				}
+			})
+
+			It("should return an error without any Options", func() {
+				opts := []Option{}
+				Expect(g.Generate(cfg, opts...)).To(MatchError(genutil.InternalError("getWriter must be set")))
```

Comment:
Can we pre-define all the internal error types in genutil and just use them here and in the generator?

Right now this is akin to redefining a constant string in two places.